internal/multistorage/cache/file.go

```diff
@@ -0,0 +1,113 @@
+package cache
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"time"
+
+	"golang.org/x/sys/unix"
+)
+
+// StatusFile is stored on a disk and therefore shared between all WAL-G processes and commands.
+const StatusFile = "/tmp/.walg_storage_status_cache"
+
+type storageStatuses map[string]status
+
+type status struct {
+	Alive   bool      `json:"alive"`
+	Checked time.Time `json:"checked"`
+}
+
+func updateFileContent(oldContent storageStatuses, checkResult map[string]bool) (newContent storageStatuses) {
+	newContent = make(storageStatuses, len(oldContent))
+	for storage, status := range oldContent {
+		newContent[storage] = status
+	}
+
+	checkTime := time.Now()
+	for s, alive := range checkResult {
+		newContent[s] = status{
+			Alive:   alive,
+			Checked: checkTime,
+		}
+	}
+
+	return newContent
+}
+
+func readFile() (storageStatuses, error) {
+	file, err := os.OpenFile(StatusFile, os.O_RDONLY, 0666)
+	if os.IsNotExist(err) {
+		return nil, nil
+	}
+	if err != nil {
+		return nil, fmt.Errorf("open cache file: %w", err)
+	}
+	defer func() { _ = file.Close() }()
+
+	err = lockFile(file, false)
+	if err != nil {
+		return nil, fmt.Errorf("acquire shared lock for the cache file: %w", err)
+	}
+
+	bytes, err := io.ReadAll(file)
+	if err != nil {
+		return nil, fmt.Errorf("read cache file: %w", err)
+	}
+
+	var content storageStatuses
+	err = json.Unmarshal(bytes, &content)
+	if err != nil {
+		return nil, fmt.Errorf("unmarshal cache file content: %w", err)
+	}
+
+	return content, nil
+}
+
+func writeFile(content storageStatuses) error {
+	bytes, err := json.Marshal(content)
+	if err != nil {
+		return fmt.Errorf("marshal cache file content: %w", err)
+	}
+
+	file, err := os.OpenFile(StatusFile, os.O_RDWR|os.O_CREATE, 0666)
+	if err != nil {
+		return fmt.Errorf("open cache file: %w", err)
+	}
+	defer func() { _ = file.Close() }()
+
+	err = lockFile(file, true)
+	if err != nil {
+		return fmt.Errorf("acquire exclusive lock for the cache file: %w", err)
+	}
+
+	err = file.Truncate(int64(len(bytes)))
```

Comment:
Truncate the file here, because otherwise, it will be a problem when a bigger file is overwritten by a smaller one. It usually can be truncated by `os.O_TRUNC` flag on opening, but it's not the case here - we shouldn't change the file before locking it.

TBH, this "O_TRUNC" flag may be implemented "lazy" (doesn't change the file size before actual writing), and it should be so if the file size is stored in a file descriptor, but I didn't find any proofs after a quick glance. It may also be implementation-dependent. A separate `Truncate` call must work in any case